### PR TITLE
Fail with a friendlier error message when response has not been set

### LINKED ISSUE #382: String Reference not set to an instance of a String: Parameter name: s
I am working on a program that must upload files to various clients it has worked well except for one client.

I have stripped out the relevant code that I am using to make it easier to debug and I have included it below.

I get the error as per the subject line of this submission when I execute it, can anyone provide some clarity for me as to what  am doing wrong, I am an old programmer and have learnt to accept that it is invariably my fault:

The error occurs when it tries to execute the connect

The program is written in Visual Basic using Visual Studio 3013

Imports Renci.SshNet

Public Class Form

    Private Sub cmdConnect_Click(sender As Object, e As EventArgs) Handles cmdConnect.Click

        Dim strURL As String
        Dim strPort As Integer
        Dim strUser As String
        Dim strPassword As String
     
        strURL = "xxxxxxx"
        strPort = 22
        strUser = "xxx"
        strPassword = "XXXXXXXXXXXX"

        Try

            Dim kauth As New KeyboardInteractiveAuthenticationMethod(strUser)

            Dim pauth As New PasswordAuthenticationMethod(strUser, strPassword)

            Dim ConnNfo As New ConnectionInfo(strURL, strPort, strUser, pauth, kauth)

            Using Sftp As New SftpClient(ConnNfo)

                Sftp.Connect()

            End Using

            MsgBox("Connected")

        Catch ex As Exception

            MsgBox(ex.Message)

        End Try

    End Sub

End Class

Regards

Juliana

### LINKED ISSUE #1237: String reference not set to an instance of a String.  Parameter name: s - version 2020.0.2
Hi All

i am using version 2020.0.2 and getting the following exception while trying to connect to linux machine

System.ArgumentNullException: String reference not set to an instance of a String.

Parameter name: s

   at Renci.SshNet.KeyboardInteractiveAuthenticationMethod.Authenticate(Session session)

   at Renci.SshNet.ClientAuthentication.TryAuthenticate(ISession session, AuthenticationState authenticationState, String[] allowedAuthenticationMethods, SshAuthenticationException& authenticationException)

   at Renci.SshNet.ClientAuthentication.Authenticate(IConnectionInfoInternal connectionInfo, ISession session)

   at Renci.SshNet.ConnectionInfo.Authenticate(ISession session, IServiceFactory serviceFactory)

   at Renci.SshNet.Session.Connect()

   at Renci.SshNet.BaseClient.CreateAndConnectSession()

   at Renci.SshNet.BaseClient.Connect()

when trying to connect vis Putty everything works just fine.

` var keyboardAuth = new KeyboardInteractiveAuthenticationMethod(username);
                keyboardAuth.AuthenticationPrompt += keyboardAuthentication_AuthenticationPrompt;

                var passwordAuth = new Renci.SshNet.PasswordAuthenticationMethod(username, password.GetStringFromSecureString());

                var connectionInfo = new Renci.SshNet.ConnectionInfo(machine.Address, port, username, keyboardAuth, passwordAuth);
                connectionInfo.Timeout = connectionTimeout;

                _client = new Renci.SshNet.SshClient(connectionInfo);
_client.Connect()`

could you please explain what is happening here?

### PR DESCRIPTION
**Pull request by Rob-Hague**
...in keyboard-interactive mode.

Closes #1237 
Closes #382

### CODE REVIEW

**Comment by WojciechNagorski** [comment]
Is there a possibility to add a test for that?

**Comment by Rob-Hague** [comment]
I wrote a unit test, but there were several issues with using a mocked `ISession` relating to visibility.

I felt that an integration test was a bit heavy for this, but I'll look at adding one anyway.

File: test/Renci.SshNet.IntegrationTests/AuthenticationTests.cs [hunk 1]
```diff
@@ -423,5 +423,35 @@ public void KeyboardInteractiveConnectionInfo()
             Assert.AreEqual(connectionInfo.Host, SshServerHostName);
             Assert.AreEqual(connectionInfo.Username, User.UserName);
         }
+
+        [TestMethod]
+        public void KeyboardInteractive_NoReponseSet_ThrowsSshAuthenticationException()
```
Comment (Rob-Hague) [inline test/Renci.SshNet.IntegrationTests/AuthenticationTests.cs]:
```suggestion
        public void KeyboardInteractive_NoResponseSet_ThrowsSshAuthenticationException()
```

File: src/Renci.SshNet/KeyboardInteractiveAuthenticationMethod.cs [hunk 2]
```diff
@@ -125,9 +124,19 @@ private void Session_UserAuthenticationInformationRequestReceived(object sender,
 
                         var informationResponse = new InformationResponseMessage();
 
-                        foreach (var response in from r in eventArgs.Prompts orderby r.Id ascending select r.Response)
```
Comment (WojciechNagorski) [inline src/Renci.SshNet/KeyboardInteractiveAuthenticationMethod.cs:128]: Why are you removing ordering by `r.Id`? Is that intentional??

Comment (Rob-Hague) [reply src/Renci.SshNet/KeyboardInteractiveAuthenticationMethod.cs:128]: The `Prompts` collection is read-only and the order never changes - it is redundant

**Comment by scott-xu** [comment]
What about separate the two


– Response null check
- IReadOnlyList and order removal

**Comment by Rob-Hague** [comment]
I'll just add the order by back. There is no need to waste time on this

File: src/Renci.SshNet/KeyboardInteractiveAuthenticationMethod.cs [hunk 3]
```diff
@@ -125,9 +124,19 @@ private void Session_UserAuthenticationInformationRequestReceived(object sender,
...
+                        foreach (var prompt in eventArgs.Prompts)
```
Comment (Rob-Hague) [inline src/Renci.SshNet/KeyboardInteractiveAuthenticationMethod.cs]:
```suggestion
                        foreach (var prompt in eventArgs.Prompts)
                        foreach (var prompt in eventArgs.Prompts.OrderBy(r => r.Id))
```

**Review by drieseng — APPROVED** [review]

**Review by WojciechNagorski — APPROVED** [review]

**Comment by WojciechNagorski** [comment]
The 2023.0.1 version has been released to Nuget: https://www.nuget.org/packages/SSH.NET/2023.0.1

**Pull request merged**